https://bugs.gentoo.org/128044
http://gcc.gnu.org/PR26969

2006-10-19 Ira Rosen  <irar@il.ibm.com>

	Backport from mainline:
	2006-08-07  Victor Kaplansky <victork@il.ibm.com>

	PR tree-optimization/26969
	* tree-vect-analyze.c (vect_analyze_loop_form): Add check of latch
	with an empty list of PHIs.

--- branches/gcc-4_1-branch/gcc/tree-vect-analyze.c	2006/10/19 09:19:21	117882
+++ branches/gcc-4_1-branch/gcc/tree-vect-analyze.c	2006/10/19 11:18:25	117883
@@ -1878,7 +1878,8 @@
      that the loop is represented as a do-while (with a proper if-guard
      before the loop if needed), where the loop header contains all the
      executable statements, and the latch is empty.  */
-  if (!empty_block_p (loop->latch))
+  if (!empty_block_p (loop->latch)
+        || phi_nodes (loop->latch))
     {
       if (vect_print_dump_info (REPORT_BAD_FORM_LOOPS))
         fprintf (vect_dump, "not vectorized: unexpected loop form.");
